upki-fetch/src/main.rs

```diff
@@ -0,0 +1,325 @@
+    let args = Args::parse();
+    if args.verbose {
+        env_logger::builder()
+            .filter_level(log::LevelFilter::Trace)
+            .init();
+    }
+
+    let local = match args.cache_dir {
+        Some(dir) => dir,
+        None => directories::ProjectDirs::from("dev", "rustls", "upki")
+            .ok_or_else(|| Report::msg("cannot determine home directory"))?
+            .data_local_dir()
+            .to_owned(),
+    };
+    log::info!("Local directory is {local:?}");
+
+    match args.command {
+        Command::Fetch { url } => fetch(&local, &url).await?,
+        Command::Verify => verify(&local)?,
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, Parser)]
+#[command(author, version, about)]
+struct Args {
+    #[command(subcommand)]
+    command: Command,
+
+    /// Where to find and save the local files.
+    ///
+    /// If not specified, this tool uses a platform-standard, local data directory.
+    cache_dir: Option<PathBuf>,
+
+    /// Emit logging output.
+    #[arg(long)]
+    verbose: bool,
+}
+
+#[derive(Debug, Subcommand)]
+enum Command {
+    /// Update the local cache of crlite filters, downloading them as needed.
+    ///
+    /// If the `--cache-dir` path does not exist, this tool creates it and parent directories.
+    ///
+    /// This also deletes filters that become unreferenced.
+    Fetch {
+        /// The remote URL.
+        #[arg(default_value = "https://upki.rustls.dev/")]
+        url: String,
+    },
+    /// Verifies the local cache of crlite filters.
+    ///
+    /// Exits zero if the manifest and all referenced filters are present.
+    ///
+    /// Exits non-zero if the manifest if any filter file is missing or
+    /// corrupt.
+    ///
+    /// This command does no network IO.  It does not say anything whether the files are up-to-date or recent.
+    Verify,
+}
+
+async fn fetch(local: &Path, remote_url: &str) -> Result<(), Report> {
+    log::info!("Fetching {remote_url} into {local:?}...");
+
+    let client = reqwest::Client::builder()
+        .use_rustls_tls()
+        .timeout(Duration::from_secs(REQUEST_TIMEOUT))
+        .user_agent(format!(
+            "{} v{} ({})",
+            env!("CARGO_PKG_NAME"),
+            env!("CARGO_PKG_VERSION"),
+            env!("CARGO_PKG_REPOSITORY")
+        ))
+        .build()
+        .wrap_err("failed to create HTTP client")?;
+
+    let manifest_url = format!("{}{}", remote_url, MANIFEST_JSON);
+    let response = client
+        .get(&manifest_url)
+        .send()
+        .await
+        .wrap_err("failed to fetch manifest")?
+        .error_for_status()
+        .wrap_err("HTTP error while fetching manifest")?;
+
+    let manifest: Manifest = response
+        .json()
+        .await
+        .wrap_err("failed to parse manifest JSON")?;
+
+    introduce_manifest(&manifest)?;
+
+    fs::create_dir_all(local).wrap_err("failed to create local directory")?;
+    let plan = form_plan(&manifest, remote_url, local)?;
```

Comment:
At this stage, the plan/step abstraction feels like it's over-engineered for what we need to do (basically, AIUI, download a bunch of files and unlink unused files). Is there a near future where this gets complex and the indirection starts being useful?

That is, I could see this being just `downloads` (a `Vec<Download>`) and the existing `unreferenced_files` (potentially renamed).